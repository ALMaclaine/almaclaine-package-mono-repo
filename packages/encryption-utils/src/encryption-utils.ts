import {hash, compare} from 'bcrypt';

export async function encryptPassword(password: string, rounds: number = 64000) {
    return hash(password, rounds);
}

export async function comparePassword(providedPassword: string, actualPassword: string) {
    return compare(providedPassword, actualPassword);
}
