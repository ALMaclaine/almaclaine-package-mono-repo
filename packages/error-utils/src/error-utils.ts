export class ALMError extends Error {
    constructor(message: string, public readonly type: string) {
        super(message);
    }
}
