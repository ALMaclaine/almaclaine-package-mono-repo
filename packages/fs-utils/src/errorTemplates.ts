export const doesNotExistError = (path: string): string => `Provided path ${path} does not exist: ${path}.`;

export const alreadyExistsError = (path: string): string => `Provided file/directory already exists: ${path}.`;

export const isNotDirectoryError = (path: string): string => `Provided path ${path} is not a directory.`;

export const isDirectoryError = (path: string): string => `Provided path ${path} is a directory.`;

export const isNotJsonFilError = (path: string): string => `Provided path ${path} is not a json file. Extension must be .json`;

