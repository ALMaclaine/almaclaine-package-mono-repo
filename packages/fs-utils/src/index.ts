export * from './fs-utils';
export * from './errorTemplates';
