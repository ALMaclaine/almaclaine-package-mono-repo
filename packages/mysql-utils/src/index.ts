export * from './mysql-utils';
export type {ConnectionInfo} from './types';
