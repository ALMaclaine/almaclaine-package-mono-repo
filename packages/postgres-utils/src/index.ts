export * from './postgres-utils';
export * from './ConnectionManager';
