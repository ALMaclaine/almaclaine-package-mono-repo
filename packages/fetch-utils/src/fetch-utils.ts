const fetch = require('cross-fetch');
import {ALMError} from "@almaclaine/error-utils";

export const ErrorTypes = {
    FetchFailed: 'FetchFailed'
}

export class FetchFailedError extends ALMError {
    constructor(message: string) {
        super(message, ErrorTypes.FetchFailed);
    }
}

export async function almFetch(url: string, options: {[key: string]: any} = {}) {
    const res = await fetch(url, options);
    if(!res.ok) {
        throw new FetchFailedError(`Fetch failed with status code: ${res.status}`);
    }
    return res;
}
