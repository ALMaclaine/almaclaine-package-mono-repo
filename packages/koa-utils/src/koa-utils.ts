import {validateEmail as validateEmailBoolean, InvalidEmail} from '@almaclaine/general-utils';
import {ALMError} from "@almaclaine/error-utils";
import {appendFileSync} from "fs";

export const ErrorTypes = {
    MissingQueryParameter: 'MissingQueryParameter',
    MissingBodyProperty: 'MissingBodyProperty'
}

export class MissingQueryParameterError extends ALMError {
    constructor(message: string) {
        super(message, ErrorTypes.MissingQueryParameter);
    }
}

export class MissingBodyPropertyError extends ALMError {
    constructor(message: string) {
        super(message, ErrorTypes.MissingBodyProperty);
    }
}

export function errorHandler(errorSet: Set<String>, unknownErrorLogFilePath: string, knownErrorLogFilePath: string) {
    return async (ctx, next) => {
        try {
            await next();
        } catch (error) {
            const {error: errorType, message} = error;
            if (errorSet.has(errorType)) {
                console.error(errorType);
                ctx.status = 400;
                ctx.body = {error: message}
            } else {
                ctx.status = 500;
                let errorText;
                errorText = `${new Date()}\n`;
                errorText += error + '\n';
                appendFileSync(unknownErrorLogFilePath, errorText);
                ctx.body = {error: 'Unknown error occurred'};
            }

            let errorText;
            if (error.error) {
                errorText = `${new Date()}\n`;
                errorText += 'Error:\n';
                errorText += error.error + '\n';
                errorText += 'Stack Trace:\n';
                if (error.stacktrace) errorText += error.stacktrace.stack + '\n\n';
            } else {
                errorText = error.stack;
            }
            appendFileSync(knownErrorLogFilePath, errorText);
        }
    }
}

function _allKeysInArray(arr, keys) {
    const values = [];

    for(const key of keys) {
        if(!(key in arr)) {
            values.push(key);
        }
    }

    return {
        truth: values.length === 0,
        values
    }
}

function _anyKeyInArray(arr, keys) {
    return {
        truth: keys.some(e => e in arr),
        values: keys
    }
}

export function validateQueryParamsAll(params) {
    return async function validateParams(ctx, next) {
        const {truth, values} = _allKeysInArray(ctx.request.query, params);
        if(!truth) {
            throw new MissingQueryParameterError(`Missing Query Parameters: ${values.join(', ')}`);
        }
        await next();
    }
}

export function validateQueryParamsAny(params) {
    return async function validateParams(ctx, next) {
        const {truth, values} = _anyKeyInArray(ctx.request.query, params);
        if(!truth) {
            throw new MissingQueryParameterError(`Missing Query Parameters, Must have one of: ${values.join(', ')}`);
        }
        await next();
    }
}

export function validateBodyProps(props) {
    return async function validateProps(ctx, next) {
        const {truth, values} = _allKeysInArray(ctx.request.query, props);
        if(!truth) {
            throw new MissingBodyPropertyError(`Missing Body Properties: ${values.join(', ')}`);
        }
        await next();
    }
}

export function validateEmail(email) {
    if (!validateEmailBoolean(email)) {
        throw new InvalidEmail(`Invalid Email: Must provide valid email.`)
    }
}
