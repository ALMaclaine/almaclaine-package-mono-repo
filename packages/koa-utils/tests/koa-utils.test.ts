import * as koaUitls from '../src';

describe('koa-utils', () => {
    it('Has Correct API', () => {
        const keys = Object.keys(koaUitls);
        expect(keys.length).toBe(9);
        expect(keys).toMatchSnapshot();
    });
});
