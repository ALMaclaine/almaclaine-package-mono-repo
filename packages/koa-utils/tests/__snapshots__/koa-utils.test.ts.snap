// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`koa-utils Has Correct API 1`] = `
Array [
  "ErrorTypes",
  "MissingQueryParameterError",
  "MissingBodyProperty",
  "InvalidEmail",
  "errorHandler",
  "validateQueryParamsAll",
  "validateQueryParamsAny",
  "validateBodyProps",
  "validateEmail",
]
`;
